File: backend/app/models/credit_model.py
from datetime import date, datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


CreditStatus = Literal["pendiente", "parcial", "pagado", "vencido", "cancelado"]


class CreditCreate(BaseModel):
    customer_id: int
    concepto: str = Field(..., min_length=3)
    monto_total: float = Field(..., gt=0)
    fecha_limite: Optional[date] = None
    notas: Optional[str] = None
    usuario: Optional[str] = "admin"


class CreditUpdate(BaseModel):
    concepto: Optional[str] = None
    fecha_limite: Optional[date] = None
    notas: Optional[str] = None
    estado: Optional[CreditStatus] = None


class CreditAdmin(BaseModel):
    credit_id: int
    customer_id: int
    customer_nombre: str
    concepto: str
    monto_total: float
    monto_pagado: float
    saldo_pendiente: float
    fecha_limite: Optional[date] = None
    estado: CreditStatus
    notas: Optional[str] = None
    usuario: str
    activo: bool
    created_at: datetime
    updated_at: datetime


class CreditPaymentCreate(BaseModel):
    monto: float = Field(..., gt=0)
    metodo_pago: Literal["efectivo", "transferencia", "tarjeta", "otro"] = "efectivo"
    nota: Optional[str] = None
    usuario: Optional[str] = "admin"


class CreditPayment(BaseModel):
    payment_id: str
    credit_id: int
    customer_id: int
    customer_nombre: str
    monto: float
    metodo_pago: str
    nota: Optional[str] = None
    usuario: str
    created_at: datetime


class CreditPaymentsResponse(BaseModel):
    total: int
    pagos: list[CreditPayment]


class ApiMessage(BaseModel):
    message: str
    customer_id: Optional[int] = None
    credit_id: Optional[int] = None

File: cash-control-app/lib/modules/business/business/models/credit_model.dart
class CreditModel {
  final int creditId;
  final int customerId;
  final String customerNombre;
  final String concepto;
  final double montoTotal;
  final double montoPagado;
  final double saldoPendiente;
  final String estado;
  final String? fechaLimite;
  final String? notas;
  final String usuario;
  final bool activo;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const CreditModel({
    required this.creditId,
    required this.customerId,
    required this.customerNombre,
    required this.concepto,
    required this.montoTotal,
    required this.montoPagado,
    required this.saldoPendiente,
    required this.estado,
    this.fechaLimite,
    this.notas,
    required this.usuario,
    required this.activo,
    this.createdAt,
    this.updatedAt,
  });

  factory CreditModel.fromJson(Map<String, dynamic> json) {
    return CreditModel(
      creditId: _toInt(json['credit_id']),
      customerId: _toInt(json['customer_id']),
      customerNombre: json['customer_nombre']?.toString() ?? '',
      concepto: json['concepto']?.toString() ?? '',
      montoTotal: _toDouble(json['monto_total']),
      montoPagado: _toDouble(json['monto_pagado']),
      saldoPendiente: _toDouble(json['saldo_pendiente']),
      estado: json['estado']?.toString() ?? 'pendiente',
      fechaLimite: json['fecha_limite']?.toString(),
      notas: json['notas']?.toString(),
      usuario: json['usuario']?.toString() ?? '',
      activo: json['activo'] == true,
      createdAt: _toDateTime(json['created_at']),
      updatedAt: _toDateTime(json['updated_at']),
    );
  }

  Map<String, dynamic> toCreateJson() {
    return {
      'customer_id': customerId,
      'concepto': concepto,
      'monto_total': montoTotal,
      'fecha_limite': fechaLimite,
      'notas': notas,
      'usuario': usuario,
    };
  }

  static int _toInt(dynamic value) {
    if (value is int) return value;
    if (value is double) return value.toInt();
    if (value is String) return int.tryParse(value) ?? 0;
    return 0;
  }

  static double _toDouble(dynamic value) {
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? 0.0;
    return 0.0;
  }

  static DateTime? _toDateTime(dynamic value) {
    if (value == null) return null;
    return DateTime.tryParse(value.toString());
  }
}
